core/client/src/block_builder/block_builder.rs

```diff
@@ -97,13 +110,30 @@ where
 
 	/// Consume the builder to return a valid `Block` containing all pushed extrinsics.
 	pub fn bake(mut self) -> error::Result<Block> {
-		self.header = self.api.finalize_block_with_context(&self.block_id, ExecutionContext::BlockConstruction)?;
+		self.bake_impl()?;
+		Ok(<Block as BlockT>::new(self.header, self.extrinsics))
+	}
+
+	fn bake_impl(&mut self) -> error::Result<()> {
+		self.header = self.api.finalize_block_with_context(
+			&self.block_id, ExecutionContext::BlockConstruction
+		)?;
 
 		debug_assert_eq!(
 			self.header.extrinsics_root().clone(),
-			HashFor::<Block>::ordered_trie_root(self.extrinsics.iter().map(Encode::encode)),
+			HashFor::<Block>::ordered_trie_root(
+				self.extrinsics.iter().map(Encode::encode)
+			),
 		);
+		Ok(())
+	}
 
-		Ok(<Block as BlockT>::new(self.header, self.extrinsics))
+	/// Consume the builder to return a valid `Block` containing all pushed extrinsics
+	/// and the generated proof.
+	pub fn bake_and_extract_proof(mut self) -> error::Result<(Block, Option<Vec<Vec<u8>>>)> {
+		self.bake_impl()?;
+
+		let proof = self.api.extract_proof();
```

Comment:
What we need to be careful of is that the nodes used when finding the storage root are also included. I think this is the case because `storage_root` is called from the runtime, but could use some confirmation.

Reply:
(disregard)